model/riscv_vmem.sail

```diff
@@ -0,0 +1,855 @@
+  vpn_size_bits       : {10,  9},        //   10     9     9
+
+  // PTE
+  levels              : { 2,  3,  4},    //    2     3     4
+  log_pte_size_bytes  : { 2,  3},        //    2     3     3
+  pte_msbs_lsb_index  : {32, 54},        //   32    54    54
+  pte_msbs_size_bits  : { 0, 10},        //    0    10    10
+  pte_PPNs_lsb_index  : {10},            //   10    10    10
+  pte_PPNs_size_bits  : {22, 44},        //   22    44    44
+  pte_PPN_j_size_bits : {10,  9}         //   10     9     9
+}
+
+// PRIVATE
+let sv32_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 9,
+       satp_asid_lsb_index = 22,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 22,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 32,
+       vpn_size_bits       = 10,
+
+       // PTE
+       levels              = 2,
+       log_pte_size_bytes  = 2,    // 4 Bytes
+       pte_msbs_lsb_index  = 32,
+       pte_msbs_size_bits  = 0,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 22,
+       pte_PPN_j_size_bits = 10
+}
+
+// PRIVATE
+let sv39_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 39,
+       vpn_size_bits       = 9,
+
+       // PTE
+       levels              = 3,
+       log_pte_size_bytes  = 3,    // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+
+// PRIVATE
+let sv48_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 48,
+       vpn_size_bits       = 9,
+
+       // PTE
+       levels              = 4,
+       log_pte_size_bytes  = 3,    // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+
+// For future; not currently used
+// PRIVATE
+/*
+let sv57_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 57,
+       vpn_size_bits       = 9,
+
+       // PTE
+       levels              = 5,
+       log_pte_size_bytes  = 3,    // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+*/
+
+// PRIVATE
+val      undefined_SV_Params : unit -> SV_Params
+
+function undefined_SV_Params() = sv32_params
```

Comment:
Currently you need it to work around an issue where Sail tries to figure out how it could do `let x : SV_Params = undefined` even though the code never does this 😞 

I've fixed Sail so it's stricter about what can be undefined in https://github.com/rems-project/sail/commit/ff2647a35c9e65df38d7e8981d54bb64de443c95, but would need a new release before we could rely on that.

Reply:
Ok, I’ve added a comment for this.